### ISSUE TITLE
Different result for `WITH q AS (SELECT 'foo' AS x) SELECT x, pg_typeof(x) FROM q;` on GPDB 6 vs GPDB 7

### ISSUE THREAD

**Issue by Chibin** [greenplum-db/gpdb#15796]
### Greenplum version or build
GPDB6:
PostgreSQL 9.4.26 (Greenplum Database 6.23.4 build dev)
GPDB7:
PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.3+dev.261.gb5ec0fb80ce build dev)

### Expected behavior
PG12 and GPDB7
```
postgres=# WITH q AS (SELECT 'foo' AS x) SELECT x, pg_typeof(x) FROM q;
  x  | pg_typeof
-----+-----------
 foo | text
(1 row)
```
### Actual behavior
GPDB6
```
postgres=# WITH q AS (SELECT 'foo' AS x) SELECT x, pg_typeof(x) FROM q;
  x  | pg_typeof
-----+-----------
 foo | unknown
(1 row)
```

Side note: If you do just the following, they all have the same result:
```
postgres=# select pg_typeof('foo');
 pg_typeof
-----------
 unknown
(1 row)
```

**Comment by kainwen** [comment]
Postgres 9.4 also has the same behavior as GPDB6:

```
psql (9.4.26)
Type "help" for help.

gpadmin=# WITH q AS (SELECT 'foo' AS x) SELECT x, pg_typeof(x) FROM q;
  x  | pg_typeof
-----+-----------
 foo | unknown
(1 row)

gpadmin=# select version();
                                                   version
--------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.4.26 on x86_64-unknown-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit
(1 row)

gpadmin=#
```

So this is an upstream issue and Postgres later fixes it (not backport to 9.4).


---------

I think we have decided that not to change 6X behavor.


Refer to:

https://github.com/greenplum-db/gpdb/pull/9655
and
https://github.com/greenplum-db/gpdb/pull/9686

**Comment by kainwen** [comment]
Close this, see comments above.
